File: 34.py
// Time Complexity : O(log n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class Solution(object): 

#### Main function

    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        result = [-1,-1] 
        
        result[0] = self.findStartingIndex(nums, target)   
        result[1] = self.findEndingIndex(nums, target) 
        
        return result 
		
#### Helper function 1 

    def findStartingIndex(self, nums, target):
        index = -1 
        low, high = 0, len(nums) -1 
        
        while low <= high: 
            mid = low + (high - low)//2 
			
            if nums[mid] == target:
                index = mid 
                high = mid - 1 
            elif nums[mid] > target:  
                high = mid - 1 
            else:  
                low = mid + 1        
        
        return index
        
#### Helper function 2
  
    def findEndingIndex(self, nums, target):
        index = -1
        low, high = 0, len(nums) -1
        
        while low <= high:
            
            mid = low + (high - low)//2
            
            if nums[mid] == target:
                index = mid
                low = mid + 1 
            elif nums[mid] > target: 
                high = mid - 1
            else:
                 low = mid + 1
            
            
        
        return index
        
        

File: 34.java
// Time Complexity : O(log n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : taken algorithm from youtube
//



// Your code here along with comments explaining your approach
//used modified binary search twice for start index and end index. for first index, when element is found at mid go to left ... and go to right for last index

class Solution {
    public int[] searchRange(int[] nums, int target) {
        
        int start=0;
        int end=nums.length-1;
        int ans[]={-1,-1};
        
        while(start<=end){
           int mid=((end-start)/2)+start;
            
            if(nums[mid]==target){
                ans[0]=mid;
                end=mid-1;
            }else if(target<nums[mid]){
                end=mid-1;
            }else if(target>nums[mid]){
                start=mid+1;
            }
        }
        
         
        int start1=0;
        int end1=nums.length-1;
        
        while(start1<=end1){
           int mid1=((end1-start1)/2)+start1;
            
            if(nums[mid1]==target){
                ans[1]=mid1;
                start1=mid1+1;
            }else if(target<nums[mid1]){
                end1=mid1-1;
            }else if(target>nums[mid1]){
                start1=mid1+1;
            }
        }
        return ans;
    }
}